{"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"Dh0g7Jtj3ztN"},"outputs":[],"source":["import pandas as pd\n","import os\n","from scipy.io import mmread\n","import numpy as np\n","from scipy.sparse import csr_matrix\n","import matplotlib.pyplot as plt\n","from tqdm import tqdm\n","import math"]},{"cell_type":"markdown","metadata":{"id":"oOncIfFY30vd"},"source":["All sequencing data used in this study was obtained from the Gene Expression Omnibus repository under the accession code GSE219281."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"b0su_0eu3ztP"},"outputs":[],"source":["rna_folder_path = \"/Users/elizabethchang/Library/CloudStorage/GoogleDrive-ec3055@columbia.edu/My Drive/final projects CBMFW4761 BINFG4002/data/c9als data/GSE219281_RAW/raw snRNA motor cortex\"\n","atac_folder_path = \"/Users/elizabethchang/Library/CloudStorage/GoogleDrive-ec3055@columbia.edu/My Drive/final projects CBMFW4761 BINFG4002/data/c9als data/GSE219281_RAW/raw snATAC motor cortex\"\n","\n","patients = ['A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'C1', 'C2', 'C3', 'C4', 'C5', 'C6']\n","\n","cell_types = ['Astro', 'Exc']\n","\n","rna_files = {}\n","atac_files = {}\n","\n","for patient in patients:\n","    rna_files[patient] = [\"matrix\", \"features\", \"barcodes\"]\n","\n","rna_metadata = pd.read_csv(\"snRNA metadata.csv\")\n","rna_metadata = rna_metadata[rna_metadata[\"annotation_major_cell_type\"].isin(cell_types)]\n","\n","atac_metadata = pd.read_csv(\"snATAC metadata.csv\")\n","atac_metadata = atac_metadata[atac_metadata[\"annotation_major_cell_type\"].isin(cell_types)]\n","\n","genes = list(pd.read_csv(\"TF_target_binary_matrix.csv\")[\"gene\"])"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"j7IisnfQ3ztQ"},"outputs":[],"source":["for file in os.listdir(rna_folder_path):\n","    split_string = file.split(\"_\")\n","    if len(split_string) > 3:\n","        file_type = split_string[-1]\n","        patient = split_string[3]\n","        file_path = os.path.join(rna_folder_path, file)\n","\n","        if file_type == \"features.tsv\":\n","            rna_files[patient][1] = file_path\n","        elif file_type == \"matrix.mtx\":\n","            rna_files[patient][0] = file_path\n","        elif file_type == \"barcodes.tsv\":\n","            rna_files[patient][2] = file_path\n","\n","for file in os.listdir(atac_folder_path):\n","    split_string = file.split(\"_\")\n","    if len(split_string) > 3:\n","        cohort = split_string[2]\n","        patient = split_string[3]\n","        file_type = split_string[-1]\n","        file_path = os.path.join(atac_folder_path, file)\n","        if file_type == \"fragments.tsv\":\n","            atac_files[patient] = file_path\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"RQAfrF813ztQ","outputId":"dfed167a-a440-483f-9bed-1ea51d8d2e3d"},"outputs":[{"name":"stderr","output_type":"stream","text":["100%|██████████| 12/12 [46:11<00:00, 230.94s/it]\n"]}],"source":["window = 1000\n","\n","for patient in tqdm(patients):\n","    atac_file = atac_files[patient]\n","\n","    atac_df = pd.read_csv(atac_file, sep=\"\\t\", header = None)\n","    atac_df.columns = [\"peak\", \"start\", \"end\", \"cell_barcode\", \"count\"]\n","    atac_df = atac_df[~atac_df[\"peak\"].isin([\"chrX\", \"chrY\"])]\n","    patient_atac_metadata = atac_metadata[atac_metadata[\"donor_id\"] == patient][[\"cell_barcode\",\"annotation_major_cell_type\"]]\n","    atac_df = pd.merge(patient_atac_metadata, atac_df, on=\"cell_barcode\")\n","\n","    atac_df['start'] = (atac_df['start'] // window) * window\n","    atac_df['end'] = (atac_df['end'] // window) * window\n","    atac_df['end'] = atac_df.apply(lambda row: row['end'] + window if row['end'] == row['start'] else row['end'], axis=1)\n","    atac_df[\"peak\"] = (atac_df[\"peak\"] + '_' + atac_df[\"start\"].astype(str)+ '_' + atac_df[\"end\"].astype(str))\n","    atac_df.drop(columns=[\"start\", \"end\"], inplace=True)\n","\n","    features_file = rna_files[patient][1]\n","    matrix_file = rna_files[patient][0]\n","    barcodes_file = rna_files[patient][2]\n","\n","    features = pd.read_csv(features_file, sep=\"\\t\", header=None)\n","    barcodes = pd.read_csv(barcodes_file, sep=\"\\t\", header=None)\n","    sc_data = mmread(matrix_file)\n","    csr_data = csr_matrix(sc_data).transpose()\n","    matrix_df = pd.DataFrame.sparse.from_spmatrix(csr_data)\n","    matrix_df.columns = features.iloc[:, 1].tolist()\n","\n","    rna_df = pd.merge(barcodes, matrix_df, left_index=True,right_index=True)\n","    rna_df.columns.values[0] = \"cell_barcode\"\n","\n","    patient_rna_metadata = rna_metadata[rna_metadata[\"donor_id\"] == patient]\n","    rna_df = pd.merge(patient_rna_metadata[[\"cell_barcode\", \"annotation_major_cell_type\"]], rna_df, on=\"cell_barcode\" )\n","\n","    for cell in cell_types:\n","        cell_atac_df = atac_df[atac_df[\"annotation_major_cell_type\"] == cell].copy()\n","        top_peaks = cell_atac_df.groupby('peak')['count'].sum().nlargest(50000).index\n","        cell_atac_df = cell_atac_df[cell_atac_df['peak'].isin(top_peaks)]\n","        atac_barcode_mapping = {barcode: i+1 for i, barcode in enumerate(cell_atac_df['cell_barcode'].unique())}\n","        cell_atac_df['cell_barcode'] = cell_atac_df['cell_barcode'].map(atac_barcode_mapping)\n","        cell_atac_df['cell_id'] = cell + '_' + cell_atac_df['cell_barcode'].astype(str)\n","        cell_atac_df.drop(columns=['cell_barcode','annotation_major_cell_type'], inplace=True)\n","\n","        atac_pivot = cell_atac_df.pivot_table(index='peak', columns='cell_id', values='count', fill_value=0)\n","        atac_pivot.reset_index(inplace=True)\n","        atac_pivot.to_csv(f\"output/{patient}_{cell}_ATAC.csv\", index=False)\n","\n","        cell_rna_df = rna_df[rna_df[\"annotation_major_cell_type\"] == cell].copy()\n","        rna_barcode_mapping = {barcode: i+1 for i, barcode in enumerate(cell_rna_df['cell_barcode'].unique())}\n","        cell_rna_df['cell_barcode'] = cell_rna_df['cell_barcode'].map(rna_barcode_mapping)\n","        cell_rna_df['cell_id'] = cell + '_' + cell_rna_df['cell_barcode'].astype(str)\n","        rna_transposed_df = cell_rna_df.set_index(cell_rna_df.columns[-1]).T.reset_index()\n","        rna_transposed_df.rename(columns={'index': 'gene'}, inplace=True)\n","        rna_transposed_df = rna_transposed_df[rna_transposed_df['gene'].isin(genes)]\n","        rna_transposed_df.to_csv(f\"output/{patient}_{cell}_RNA.csv\", index=False)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"wZ_4r2Hm3ztR","outputId":"320caaa3-06b7-45c8-97d0-c7a7a9f66c4f"},"outputs":[{"name":"stdout","output_type":"stream","text":["A1_Astro_ATAC.csv: (50000, 201)\n","A1_Astro_RNA.csv: (17197, 911)\n","A1_Exc_ATAC.csv: (50000, 198)\n","A1_Exc_RNA.csv: (17197, 533)\n","A2_Astro_ATAC.csv: (50000, 235)\n","A2_Astro_RNA.csv: (17197, 352)\n","A2_Exc_ATAC.csv: (50000, 1244)\n","A2_Exc_RNA.csv: (17197, 866)\n","A3_Astro_ATAC.csv: (50000, 246)\n","A3_Astro_RNA.csv: (17197, 474)\n","A3_Exc_ATAC.csv: (50000, 1014)\n","A3_Exc_RNA.csv: (17197, 1042)\n","A4_Astro_ATAC.csv: (50000, 625)\n","A4_Astro_RNA.csv: (17197, 86)\n","A4_Exc_ATAC.csv: (50000, 1136)\n","A4_Exc_RNA.csv: (17197, 78)\n","A5_Astro_ATAC.csv: (50000, 237)\n","A5_Astro_RNA.csv: (17197, 454)\n","A5_Exc_ATAC.csv: (50000, 446)\n","A5_Exc_RNA.csv: (17197, 556)\n","A6_Astro_ATAC.csv: (50000, 154)\n","A6_Astro_RNA.csv: (17197, 629)\n","A6_Exc_ATAC.csv: (50000, 505)\n","A6_Exc_RNA.csv: (17197, 504)\n","C1_Astro_ATAC.csv: (50000, 609)\n","C1_Astro_RNA.csv: (17197, 1020)\n","C1_Exc_ATAC.csv: (50000, 1169)\n","C1_Exc_RNA.csv: (17197, 386)\n","C2_Astro_ATAC.csv: (50000, 242)\n","C2_Astro_RNA.csv: (17197, 284)\n","C2_Exc_ATAC.csv: (50000, 771)\n","C2_Exc_RNA.csv: (17197, 206)\n","C3_Astro_ATAC.csv: (50000, 218)\n","C3_Astro_RNA.csv: (17197, 519)\n","C3_Exc_ATAC.csv: (50000, 1397)\n","C3_Exc_RNA.csv: (17197, 1061)\n","C4_Astro_ATAC.csv: (50000, 303)\n","C4_Astro_RNA.csv: (17197, 530)\n","C4_Exc_ATAC.csv: (50000, 328)\n","C4_Exc_RNA.csv: (17197, 639)\n","C5_Astro_ATAC.csv: (50000, 383)\n","C5_Astro_RNA.csv: (17197, 405)\n","C5_Exc_ATAC.csv: (50000, 861)\n","C5_Exc_RNA.csv: (17197, 321)\n","C6_Astro_ATAC.csv: (50000, 267)\n","C6_Astro_RNA.csv: (17197, 491)\n","C6_Exc_ATAC.csv: (50000, 50)\n","C6_Exc_RNA.csv: (17197, 262)\n"]}],"source":["for file in sorted(os.listdir(\"output\")):\n","    df = pd.read_csv(f\"output/{file}\", encoding='unicode_escape')\n","    print(f\"{file}: {df.shape}\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"2kZBKVhC3ztS"},"outputs":[],"source":["def subset(merged_df, t, cells = 250, rows = 10000):\n","    data = \"peak\" if t==\"ATAC\" else \"gene\"\n","    row_sums = merged_df.iloc[:, 1:].sum(axis=1)\n","    col_sums = merged_df.iloc[:, 1:].sum(axis=0)\n","    merged_df[\"sum\"] = row_sums\n","    merged_df = merged_df.nlargest(columns=\"sum\", n=rows).drop(columns=\"sum\")\n","    columns_to_keep = [data] + list(col_sums.nlargest(cells).index)\n","    data_df = merged_df[columns_to_keep]\n","    data_df = pd.concat([merged_df[data],data_df.iloc[:, 1:].map(math.ceil)],axis=1)\n","    return data_df"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"7TGKrH4A3ztS","outputId":"5ea5aca5-30bc-4d21-8f52-0ff86f76ddf5"},"outputs":[{"name":"stdout","output_type":"stream","text":["merged all peaks subset/subset_Control_ATAC_Astro.csv (10000, 251)\n","merged all peaks subset/subset_ALS_ATAC_Astro.csv (10000, 251)\n","merged all peaks subset/subset_Control_RNA_Astro.csv (10000, 251)\n","merged all peaks subset/subset_ALS_RNA_Astro.csv (10000, 251)\n","merged all peaks subset/subset_Control_ATAC_Exc.csv (10000, 251)\n","merged all peaks subset/subset_ALS_ATAC_Exc.csv (10000, 251)\n","merged all peaks subset/subset_Control_RNA_Exc.csv (10000, 251)\n","merged all peaks subset/subset_ALS_RNA_Exc.csv (10000, 251)\n"]}],"source":["folder = \"output rna and atac 50000\"\n","\n","control = ['C1', 'C2', 'C3', 'C4', 'C5', 'C6']\n","als = ['A1', 'A2', 'A3', 'A4', 'A5', 'A6']\n","cohorts = [control, als]\n","cohort_names = [\"control\", \"als\"]\n","\n","astro_atac = {}\n","astro_rna = {}\n","exc_atac = {}\n","exc_rna = {}\n","\n","for file in sorted(os.listdir(folder)):\n","    if file.endswith(\".csv\"):\n","        file_path = os.path.join(folder,file)\n","        temp = file[:-4].split(\"_\")\n","        patient = temp[0]\n","        data = temp[2]\n","        cell = temp[1]\n","\n","        if cell == \"Astro\":\n","            if data == \"ATAC\":\n","                astro_atac[patient] = file_path\n","            else:\n","                astro_rna[patient] = file_path\n","        else:\n","            if data == \"ATAC\":\n","                exc_atac[patient] = file_path\n","            else:\n","                exc_rna[patient] = file_path\n","\n","dict_names = ['astro_atac', 'astro_rna', 'exc_atac', 'exc_rna']\n","files_list = [astro_atac, astro_rna, exc_atac, exc_rna]\n","\n","for file_dict, dict_name in zip(files_list, dict_names):\n","    for cohort in cohorts:\n","        df_list = []\n","        for patient in cohort:\n","            df = pd.read_csv(file_dict[patient])\n","            df_list.append(df)\n","        merged_df = None\n","        suffix_number = 0\n","        for df in df_list:\n","            if merged_df is None:\n","                merged_df = df\n","            else:\n","                suffix_number += 1\n","                suffix = '_' + str(suffix_number)\n","                merged_df = pd.merge(merged_df, df, on=\"peak\" if \"atac\" in dict_name else \"gene\", suffixes=('', suffix), how = \"outer\")\n","        c = \"Control\" if \"C1\" in cohort else \"ALS\"\n","        t = \"ATAC\" if \"atac\" in dict_name else \"RNA\"\n","        k = \"Astro\" if \"astro\" in dict_name else \"Exc\"\n","        file_name = f\"merged all peaks subset/subset_{c}_{t}_{k}.csv\"\n","        merged_df.fillna(0, inplace=True)\n","        data_df = subset(merged_df, t)\n","        data_df.to_csv(file_name, index=False)\n","        print(file_name, data_df.shape)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"0E3V7Ter3ztS","outputId":"77a38f9e-509c-40e8-ca4d-86b470c90591"},"outputs":[{"name":"stdout","output_type":"stream","text":["ALS_ATAC_Astro.csv (17741, 1693)\n","ALS_ATAC_Exc.csv (19221, 4538)\n","ALS_RNA_Astro.csv (17941, 2901)\n","ALS_RNA_Exc.csv (17941, 3574)\n","Control_ATAC_Astro.csv (18836, 2017)\n","Control_ATAC_Exc.csv (7181, 4571)\n","Control_RNA_Astro.csv (17941, 3244)\n","Control_RNA_Exc.csv (17941, 2870)\n"]}],"source":["for file in sorted(os.listdir(\"data\")):\n","    file_path = os.path.join(\"data\",file)\n","    print(file, pd.read_csv(file_path).shape)"]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.0"}},"nbformat":4,"nbformat_minor":0}
